measured.go

```diff
@@ -1,310 +1,148 @@
-			log.Errorf("Failed to report traffic data to %v: %v", reflect.TypeOf(r), err)
-		}
+func (c *conn) Read(b []byte) (int, error) {
+	c.recv.begin(mtime.Now)
+	n, err := c.Conn.Read(b)
+	c.recv.advance(n, mtime.Now())
+	if err != nil && !isTimeout(err) && err != io.EOF {
+		c.storeError(err)
 	}
+	return n, err
 }
 
-// Conn wraps any net.Conn to add statistics
-type Conn struct {
-	net.Conn
-	// arbitrary string to identify this connection, defaults to remote address
-	ID string
-	// total bytes read from this connection
-	BytesIn uint64
-	// total bytes wrote to this connection
-	BytesOut                  uint64
-	m                         *Measured
-	interval                  time.Duration
-	elapsedSinceLastSubmitted func() time.Duration
-	submitMutex               sync.Mutex
-}
-
-func (m *Measured) newConn(c net.Conn, interval time.Duration) net.Conn {
-	ra := c.RemoteAddr()
-	if ra == nil {
-		panic("nil remote address is not allowed")
+func (c *conn) Close() error {
+	if atomic.CompareAndSwapInt32(&c.closed, 0, 1) {
+		err := c.Conn.Close()
+		// Wait for tracking to finish
+		<-c.trackingFinished
+		return err
 	}
-	return &Conn{Conn: c, ID: ra.String(), m: m, interval: interval}
-}
-
-// Read() implements the function from net.Conn
-func (mc *Conn) Read(b []byte) (n int, err error) {
-	n, err = mc.Conn.Read(b)
-	atomic.AddUint64(&mc.BytesIn, uint64(n))
-	mc.submitTrafficIfNecessary()
-	return
-}
-
-// Write() implements the function from net.Conn
-func (mc *Conn) Write(b []byte) (n int, err error) {
-	n, err = mc.Conn.Write(b)
-	atomic.AddUint64(&mc.BytesOut, uint64(n))
-	mc.submitTrafficIfNecessary()
-	return
-}
-
-// Close implements the function from net.Conn
-func (mc *Conn) Close() (err error) {
-	err = mc.Conn.Close()
-	mc.submitTraffic()
-	return
+	return nil
 }
 
-func (mc *Conn) submitTrafficIfNecessary() {
-	needToSubmit := false
-	mc.submitMutex.Lock()
-	if mc.elapsedSinceLastSubmitted == nil || mc.elapsedSinceLastSubmitted() > mc.interval {
-		needToSubmit = true
-		mc.elapsedSinceLastSubmitted = mtime.Stopwatch()
+func (c *conn) storeError(err error) {
+	c.errMx.Lock()
+	if c.firstErr == nil {
+		c.firstErr = err
 	}
-	mc.submitMutex.Unlock()
-	if needToSubmit {
-		mc.submitTraffic()
-	}
-}
-
-func (mc *Conn) submitTraffic() {
-	mc.m.submitTraffic(mc.ID,
-		atomic.SwapUint64(&mc.BytesIn, 0),
-		atomic.SwapUint64(&mc.BytesOut, 0))
+	c.errMx.Unlock()
 }
 
-func (m *Measured) submitTraffic(connID string, in uint64, out uint64) {
-	if atomic.LoadInt32(&m.stopped) == 1 {
-		log.Trace("Measured stopped, not submitting traffic")
-		return
-	}
-	t := &Traffic{
-		ID:       connID,
-		BytesIn:  in,
-		BytesOut: out,
-	}
-	select {
-	case m.chTraffic <- t:
-		log.Tracef("Submitted traffic %+v", t)
-	default:
-		log.Tracef("Discarded traffic %+v", t)
-	}
+func isTimeout(err error) bool {
+	es := err.Error()
+	esl := len(es)
+	return esl >= ioTimeoutLength && es[esl-ioTimeoutLength:] == ioTimeout
```

Comment:
Yes, this is an optimization that I discovered while working on netx. The cast to net.Error is actually much more expensive than doing the string-based test.